spring-batch-samples/src/test/java/org/springframework/batch/sample/metrics/BatchMetricsTests.java

```diff
@@ -0,0 +1,168 @@
+/*
+ * Copyright 2019 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.springframework.batch.sample.metrics;
+
+import java.util.Arrays;
+import java.util.List;
+
+import io.micrometer.core.instrument.Meter;
+import io.micrometer.core.instrument.Metrics;
+import org.hamcrest.Matchers;
+import org.junit.Test;
+
+import org.springframework.batch.core.ExitStatus;
+import org.springframework.batch.core.Job;
+import org.springframework.batch.core.JobExecution;
+import org.springframework.batch.core.JobParameters;
+import org.springframework.batch.core.Step;
+import org.springframework.batch.core.configuration.annotation.EnableBatchProcessing;
+import org.springframework.batch.core.configuration.annotation.JobBuilderFactory;
+import org.springframework.batch.core.configuration.annotation.StepBuilderFactory;
+import org.springframework.batch.core.launch.JobLauncher;
+import org.springframework.batch.item.ItemReader;
+import org.springframework.batch.item.ItemWriter;
+import org.springframework.batch.item.support.ListItemReader;
+import org.springframework.batch.repeat.RepeatStatus;
+import org.springframework.context.ApplicationContext;
+import org.springframework.context.annotation.AnnotationConfigApplicationContext;
+import org.springframework.context.annotation.Bean;
+import org.springframework.context.annotation.Configuration;
+
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertThat;
+
+public class BatchMetricsTests {
+
+	@Test
+	public void testBatchMetrics() throws Exception {
+		// given
+		ApplicationContext context = new AnnotationConfigApplicationContext(MyJobConfiguration.class);
+		JobLauncher jobLauncher = context.getBean(JobLauncher.class);
+		Job job = context.getBean(Job.class);
+		
+		// when
+		JobExecution jobExecution = jobLauncher.run(job, new JobParameters());
+		
+		// then
+		assertEquals(ExitStatus.COMPLETED, jobExecution.getExitStatus());
+		List<Meter> meters = Metrics.globalRegistry.getMeters();
+		assertThat(meters, Matchers.hasSize(7));
+
+		Meter.Id itemReadMeter = meters.get(0).getId();
```

Comment:
Micrometer has a built-in testing facility for this. Use:
```
Metrics.globalRegistry.get("spring.batch.item.read")
   .tag("job.name", "job") // single tag at a time
   .tag("step.name", "step2", "status", "SUCCESS") // or you can do them all at once
   .timer();
```

If you use `get(..)` and there isn't a meter with this type/name/tags, Micrometer will throw a runtime exception and generate a detailed description of why it couldn't find it and which meters got closest to matching your requirements.

This has the added benefit of not requiring the meters to be in any certain order.

Reply:
Got it. I updated the test to get each meter with its name/tags. That's indeed a better approach to avoid getting metrics in any certain order.